src/openms/source/METADATA/SpectrumMetaDataLookup.cpp

```diff
@@ -150,40 +151,71 @@ namespace OpenMS
+                                                        const MSExperiment& exp)
+{
+    // Check if the experiment has spectra
+    if (exp.getSpectra().empty())
+    {
+        OPENMS_LOG_INFO << "No spectra found in the experiment. Skipping RT annotation." << endl;
+        return false;
+    }
+
     SpectrumLookup lookup;
+    lookup.readSpectra(exp.getSpectra());
     bool success = true;
+
+    // Iterate over peptide IDs and annotate missing RT values
     for (auto& pep : peptides)
     {
-      if (std::isnan(pep.getRT()))
-      {
-        if (lookup.empty())
-        {
-          FileHandler fh;
-          auto opts = fh.getOptions();
-          // speed up reading. We do not need the actual peaks in the spectra
-          opts.setFillData(false);
-          opts.setSkipXMLChecks(true);
-          fh.setOptions(opts);
-          fh.loadExperiment(filename, exp, {FileTypes::MZXML, FileTypes::MZML, FileTypes::MZDATA, FileTypes::MGF}, OpenMS::ProgressLogger::NONE, true, true);
-          lookup.readSpectra(exp.getSpectra());
-        }
-        String spectrum_id = pep.getSpectrumReference();
-        try
-        {
-          Size index = lookup.findByNativeID(spectrum_id);
-          pep.setRT(exp[index].getRT());
-        }
-        catch (Exception::ElementNotFound&)
+        if (std::isnan(pep.getRT())) // Only annotate peptides with missing RT
         {
-          OPENMS_LOG_ERROR << "Error: Failed to look up retention time for peptide identification with spectrum reference '" + spectrum_id + "' - no spectrum with corresponding native ID found." << endl;
-          success = false;
-          if (stop_on_error) break;
+            String native_id = pep.getSpectrumReference();
+            try
+            {
+                // Look up spectrum index by native ID and assign RT
+                Size index = lookup.findByNativeID(native_id);
+                pep.setRT(exp.getSpectra()[index].getRT());
+            }
+            catch (Exception::ElementNotFound&)
+            {
+                // Log error if the spectrum reference is not found
+                OPENMS_LOG_ERROR << "Error: Failed to look up retention time for peptide identification with spectrum reference '"
+                                 << native_id << "' - no spectrum with corresponding native ID found." << endl;
+                success = false;
+            }
         }
-      }
+    }
+
+    return success;
+}
+
+  bool SpectrumMetaDataLookup::addMissingIMToPeptideIDs(vector<PeptideIdentification>& peptides,
+                                const MSExperiment& exp)
+  { 
+    // Check if the experiment has spectra
+    if (exp.getSpectra().empty())
+    {
+        OPENMS_LOG_INFO << "No spectra found in the experiment. Skipping IM annotation." << endl;
+        return false;
+    }
+    // Check if desired IM format is present
+	if (IMTypes::determineIMFormat(exp) != IMFormat::MULTIPLE_SPECTRA)
```

Comment:
In principle we would only need IM values for the identified peptides. (e.g. identified MS2 spectra).
This check could be done in the loop.

Reply:
see suggestion below